katas/content/teleportation/Common.qs

```diff
@@ -1,7 +1,123 @@
+    // 2 - |Ψ⁺⟩ = (|01⟩ + |10⟩) / sqrt(2)
+    // 3 - |Ψ⁻⟩ = (|01⟩ - |10⟩) / sqrt(2)
+    operation StatePrep_BellState(q1 : Qubit, q2 : Qubit, state : Int) : Unit {
+        H(q1);
+        CNOT(q1, q2);
+        
+        // now we have |00⟩ + |11⟩ - modify it based on state arg
+        if state % 2 == 1 {
+            // negative phase
+            Z(q2);
+        }
+        
+        if state / 2 == 1 {
+            X(q2);
+        }
+    }
+
+    // ------------------------------------------------------
+    // Helper operation that run teleportation using the given operations to prepare the message qubit
+    // and the entangled pair, and to run sender and receiver parts of the protocol.
+    operation ComposeTeleportation(
+        bellPrepOp : ((Qubit, Qubit) => Unit), 
+        getDescriptionOp : ((Qubit, Qubit) => (Bool, Bool)), 
+        reconstructOp : ((Qubit, (Bool, Bool)) => Unit),
+        qAlice : Qubit, 
+        qBob : Qubit, 
+        qMessage : Qubit) : Unit {
+
+        bellPrepOp(qAlice, qBob);
+        let classicalBits = getDescriptionOp(qAlice, qMessage);
+        
+        // Alice sends the classical bits to Bob.
+        // Bob uses these bits to transform his part of the entangled pair into the message.
+        reconstructOp(qBob, classicalBits);
+    }
+    
+    operation SendMessage_Reference(qAlice: Qubit, qMessage: Qubit) : (Bool, Bool) {
+        CNOT(qMessage, qAlice);
+        H(qMessage);
+        return (M(qMessage) == One, M(qAlice) == One);
+    }
+
+    operation ReconstructMessage_Reference(qBob : Qubit, (b1 : Bool, b2 : Bool)) : Unit {
+        if b1 {
+            Z(qBob);
+        }
+        if b2 {
+            X(qBob);
+        }
+    }
+
+    // ------------------------------------------------------
+    // Helper operation that runs a teleportation operation (specified by teleportOp).
+    // The state to teleport is set up using an operation (specified by setupPsiOp).
+    //
+    // Specifying the state to teleport through an operation allows to get the inverse
+    // which makes testing easier.
+    operation TeleportTestHelper(
+        teleportOp : ((Qubit, Qubit, Qubit) => Unit), 
+        setupPsiOp : (Qubit => Unit is Adj)) : Bool {
+        
+        use (qMessage, qAlice, qBob) = (Qubit(), Qubit(), Qubit());
+        setupPsiOp(qMessage);
+            
+        // This should modify qBob to be identical to the state
+        // of qMessage before the function call.
+        teleportOp(qAlice, qBob, qMessage);
+            
+        // Applying the inverse of the setup operation to qBob
+        // should make it Zero.
+        Adjoint setupPsiOp(qBob);
+        if not CheckZero(qBob){
+            Message("Incorrect. The state was teleported incorrectly.");
+            DumpMachine();
+            ResetAll([qMessage, qAlice, qBob]);
+            return false;
+        }
+        ResetAll([qMessage, qAlice, qBob]);
+        Message("Correct.");
```

Comment:
We can move this message to the end of TeleportTestLoop, so that we don't print a hundred messages if teleporting |0> state succeeds and teleporting |1> fails